evm/contracts/AccessGrants.sol

```diff
@@ -25,24 +26,35 @@ contract AccessGrants {
 
     function insertGrant(
         address grantee,
-        string memory dataId,
+        string calldata dataId,
         uint256 lockedUntil
     ) external {
-        Grant memory grant = Grant({
-            owner: msg.sender,
-            grantee: grantee,
-            dataId: dataId,
-            lockedUntil: lockedUntil
-        });
-
-        bytes32 grantId = _deriveGrantId(grant);
-
-        require(_grantsById[grantId].owner == address(0), "Grant already exists");
+         _insertGrant(msg.sender, grantee, dataId, lockedUntil);
+    }
 
-        _grantsById[grantId] = grant;
-        _grantIdsByOwner[grant.owner].add(grantId);
-        _grantIdsByGrantee[grant.grantee].add(grantId);
-        _grantIdsByDataId[grant.dataId].add(grantId);
+    function insertGrantBySignature(
```

Comment:
Tangent thought: why is `insertGrant` special? Maybe we should have all operations on this contract care about owner signature, and not that the owner is the sender.

Non-actionable, I just wanted to show thoughts. I'll keep this on my radar.

Reply:
Ended up implementing this.